### PR TITLE
Fix ONNX ATen fallback for non-caffe2 engines

### LINKED ISSUE #75623: [ONNX] `ONNX_ATEN_FALLBACK` supports ONNX Runtime
Previously setting `operator_export_type=OperatorExportTypes.ONNX_ATEN_FALLBACK` produed a model that was only usable by Caffe2.
This issue tracks work to make it usable by ONNX Runtime as well.

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
This PR introduces 3 BC changes:

First, this PR propagates `BUILD_CAFFE2` flag to `libtorch` and `libtorch_python`, which is necessary for non-caffe2 ONNX runtimes when using `ONNX_ATEN_FALLBACK` operator export type.

Second, as a complement of https://github.com/pytorch/pytorch/pull/68490, this PR refactors Caffe2's Aten ops symbolics to consider not only the `operator_export_type` (aka `ONNX_ATEN_FALLBACK`) to emit Caffe2 Aten ops, but also whether `BUILD_CAFFE2` (which is called `torch.onnx._CAFFE2_ATEN_FALLBACK` in python binding) is set.

Lastly, it renames `onnx::ATen` to `aten::ATen` for ONNX spec consistency in a BC fashion.
ONNX doesn't have `ATen` op on its spec, but PyTorch ONNX converter emits them. Non-Caffe2 backend engines would be mislead by such operator's name/domain. A non-ideal workaround would be to have Aten ops handled based on its name and ignore the (non-complaint) domain. Moreover, users could incorrectly file bugs to either ONNX or ONNX Runtime when they inspect the model and notice the presence of an unspecified ONNX operator.

### CODE REVIEW

**Comment by thiagocrepaldi** [comment]
Folks, I have added you in the hope you are still the Caffe2 experts. Feel free to remove yourself and add include a colleague if that is not the case.

The proposed change aims to refactor the clever Caffe2;s Aten Op in a way that external DL frameworks (aka ONNX Runtime) can use it without having to parse `native_functions.yaml`, as external DL wont have this file shipped in and nor torch wheel apckages include it

File: test/onnx/expect/TestOperators.test_layer_norm_aten.expect [hunk 1]
```diff
@@ -111,16 +112,16 @@ graph {
         elem_type: 1
         shape {
           dim {
-            dim_value: 20
```
Comment (BowenBao) [inline test/onnx/expect/TestOperators.test_layer_norm_aten.expect:114]: Are these changes from static shape to dynamic shape intended and expected?

File: test/onnx/test_operators.py [hunk 2]
```diff
@@ -339,7 +339,8 @@ class MyModule(Module):
             def forward(self, x):
                 return MyFun.apply(x)
 
-        self.assertONNX(MyModule(), x)
+        self.assertONNX(MyModule(), x,
+                        operator_export_type=torch.onnx.OperatorExportTypes.ONNX_ATEN_FALLBACK)
```
Comment (BowenBao) [inline test/onnx/test_operators.py:343]: what happens without adding `operator_export_type`?

File: torch/CMakeLists.txt [hunk 3]
```diff
@@ -340,6 +340,12 @@ endif()
 
 
 add_library(torch_python SHARED ${TORCH_PYTHON_SRCS})
+if(DEFINED ENV{BUILD_CAFFE2})
```
Comment (BowenBao) [inline torch/CMakeLists.txt:343]: Did you add this? I'm surprised how this was not handled. Would the variable `BUILD_CAFFE2` not work without this change?

File: torch/csrc/jit/serialization/export.cpp [hunk 4]
```diff
@@ -113,13 +113,13 @@ void validateBlock(
...
               b->owningGraph()->insertNode(b->owningGraph()->create(
-                  Symbol(::c10::onnx::ATen),
+                  Symbol(::c10::aten::ATen),
                   node->inputs(),
                   node->outputs().size()));
           for (size_t i = 0; i < node->outputs().size(); ++i) {
             node->output(i)->replaceAllUsesWith(new_node->output(i));
           }
-          new_node->s_(Symbol::fromQualString("attr::operator"), "expand");
+          new_node->s_(Symbol::fromQualString("attr::operator_name"), "expand");
```
Comment (BowenBao) [inline torch/csrc/jit/serialization/export.cpp:122]: nit: attr::operator_name

File: torch/onnx/utils.py [hunk 5]
```diff
@@ -947,7 +947,7 @@ def _newNode(g, opname, outputs, *args, **kwargs):
         aten = False
         ns_opname = opname
     else:
-        aten = kwargs.pop("aten", False)
```
Comment (BowenBao) [inline torch/onnx/utils.py:950]: Is there a case where first condition validates to False, while second is True?

File: test/onnx/test_operators.py (hunk 2 above)
Comment (thiagocrepaldi) [reply test/onnx/test_operators.py:343]: `RuntimeError: ONNX export failed: Couldn't export operator aten::ATen` is raised, because a `ATen` op is created and export is on `ONNX` only mode

Comment (BowenBao) [reply test/onnx/test_operators.py:343]: isee, because previously it was onnx::aten. thx

File: torch/CMakeLists.txt (hunk 3 above)
Comment (thiagocrepaldi) [reply torch/CMakeLists.txt:343]: `BUILD_CAFFE2` exists for a long time, but it was always consumed from main `CMakeLists.txt`. 
However, we need this to be propagated to `torch/CMakeLists.txt`. I have tried everything, including the standard `if(BUILD_CAFFE2)`, but for some reason `OPTIONS()` from root `CMakeLists.txt` are not propagated to `torch/CMakeLists.txt`. It kind of make sense because I couldnt find a `add_subdirectory(torch)` on the root dir, which enables such sharing.

Finally I mimic'ed this from other places in the pytorch build script.

File: torch/csrc/jit/serialization/export.cpp (hunk 4 above)
Comment (thiagocrepaldi) [reply torch/csrc/jit/serialization/export.cpp:122]: What is the issue? I don't see the typo

Comment (BowenBao) [reply torch/csrc/jit/serialization/export.cpp:122]: using `attr::operator_name` directly?

Comment (thiagocrepaldi) [reply torch/csrc/jit/serialization/export.cpp:122]: There are 2 ways to get those attributes.

The indirect through `fromQualString("attr::operator")` and the direct using `::c10::attr::operator`. The former works for `operator` while the latter doesn't. All other attributes support both, so I renamed this for consistency in a BC way

File: torch/onnx/utils.py (hunk 5 above)
Comment (thiagocrepaldi) [reply torch/onnx/utils.py:950]: Not after unifying onnx::aten and aten::aten. Good point, removing it

File: test/onnx/expect/TestOperators.test_layer_norm_aten.expect (hunk 1 above)
Comment (thiagocrepaldi) [reply test/onnx/expect/TestOperators.test_layer_norm_aten.expect:114]: That I can't tell. It makes no sense to me considering the nature of changes that I have done. 

What I would guess is that shape inference for aten domain is somehow different than for onnx domain?

Comment (BowenBao) [reply test/onnx/expect/TestOperators.test_layer_norm_aten.expect:114]: > What I would guess is that shape inference for aten domain is somehow different than for onnx domain

Makes sense. @jiafatom should know more. It might have something to do with `TypeReliable` etc.

**Review by BowenBao — APPROVED** [review]

File: caffe2/contrib/aten/aten_op_template.h [hunk 6]
```diff
@@ -179,7 +180,8 @@ class ATenOp : public Operator<Context> {
     std::vector<std::string> attrs;
     for (const auto i : c10::irange(operator_def.arg_size())) {
       auto & attr = operator_def.arg(i);
-      if(attr.name() == "operator" || attr.name() == "type" || attr.name() == "overload_name" ) {
+      if (attr.name() == "operator_name" || attr.name() == "type" ||
```
Comment (dzhulgakov) [inline caffe2/contrib/aten/aten_op_template.h:183]: it's not backward compatible. If there are some models with serialized Caffe2 Nets that contain ATenOp it'd break. You'd need either:
1) make this op recognize both types of arguments, i.e. `operator_name` and `operator`
2) change the lowering from ONNX to Caffe2 to rename the arg back to be just `operator`

I probably prefer the second one

**Review by dzhulgakov — CHANGES_REQUESTED** [review]
What is the broader objective are you trying to achieve? Why does moving from `onnx` to `aten` namespace help you while you're still operating on TorchScript graph? Generally `aten::Aten` kind of doesn't make sense, why no just use regular `aten` ops?

**Comment by thiagocrepaldi** [comment]
> What is the broader objective are you trying to achieve? Why does moving from `onnx` to `aten` namespace help you while you're still operating on TorchScript graph? Generally `aten::Aten` kind of doesn't make sense, why no just use regular `aten` ops?

Thanks for the quick feedback Dmytro!

The broader goal is to allow other DL frameworks (e.g. Microsoft's ONNX Runtime) to execute ONNX graphs containing Caffe2's Aten nodes on graphs generated by `torch.onnx.export(...,expor_operator_type=ONNX_ATEN_FALLBACK)`. 

The renaming from `onnx::aten` to `aten::aten` was done because there is no such `ATen` operator at [ONNX spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md) and that might cause confusion for the ONNX community when they debug ONNX graphs and find those unspecified ops. `org.pytorch.aten::ATen`, on the other hand, clearly states this is a PyTorch thing as opposed to ONNX's.

The renaming from `operator` to `operator_name` was just an opportunistic improvement that I proposed given that I had compile-time errors when tried to print logs using `node->s(::c10::attr::operator)` idiom. I realized the hard way that doing `node->s(::c10::attr::name)` would work just fine but `node->s(::c10::attr::operator)` wouldn't; so the renaming fix that minor inconsistency without any side effect.

Both renaming are not strictly necessary, but IMO they add clarity (`onnx::aten` ->> `aten::ATen`) and consistency (all attributes, including the operator name, can be accessed the same way).

File: caffe2/contrib/aten/aten_op_template.h (hunk 6 above)
Comment (thiagocrepaldi) [reply caffe2/contrib/aten/aten_op_template.h:183]: I have refactored Caffe2 code to accept such change and verified using all Unit Tests available to make sure this is not BC breaking

Comment (dzhulgakov) [reply caffe2/contrib/aten/aten_op_template.h:183]: Verifying that unittests pass is not enough. If there's a serialized Caffe2 NetDef model somewhere with an ATen op inside, it won't be openable with the new code. We don't have unittest that exercises the serialized Caffe2 model afaiu.

At least at Meta we still have a few models like that laying around. So I'd prefer you not to change Caffe2's op signature. But since you have PT->ONNX->C2 two step conversion, you can evolve ONNX's operator spec as you fit while keeping the Caffe2 op spec intact for the existing models.

The similar change you pointed out is not affecting how already converted models are loaded, so it's ok.

**Comment by dzhulgakov** [comment]
> The broader goal is to allow other DL frameworks (e.g. Microsoft's ONNX Runtime) to execute ONNX graphs containing Caffe2's Aten nodes 

How are you planning to do that? By depending on Caffe2 code or by implementing similar functionality?  I wouldn’t recommend using Caffe2 code for anything, it’s really in maintenance mode and might just go away (we don’t even build with it in releases anymore).

If you want to implement something akin to Caffe2's ATenOp inside ONNXRT there are two options:
1) extending codegen
2) using dispatcher's “boxed” API, it’d be only a tiny bit slower. Example usage: https://github.com/pytorch/pytorch/blob/a9b0a921d592b328e7e80a436ef065dadda5f01b/aten/src/ATen/core/boxing/KernelFunction_test.cpp#L113 (and `call` and `callBoxed` in the same file).

Note that Caffe2's ATenOp is really old and predates the existence of these nice Dispatcher APIs.

> The renaming from onnx::aten to aten::aten was done because there is no such ATen operator at [ONNX spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md) and that might cause confusion for the ONNX community when they debug ONNX graphs and find those unspecified ops. org.pytorch.aten::ATen

Note that `onnx::` or `aten::` in this case are just nodes used inside PyTorch's graph handling. `aten::` namespace really corresponds to runnable operators, so `onnx::` prefix (or any other one) is better. You can extend export to ONNX proto to put it into whatever namespace you prefer though.

Also, you may consider just creating ONNX nodes with `org.pytorch.aten` domain and operator name being the actual op name (e.g. `conv`) instead of catch-all `Aten` operator. But that'd be a bigger change

File: caffe2/contrib/aten/aten_op_template.h (hunk 6 above)
Comment (thiagocrepaldi) [reply caffe2/contrib/aten/aten_op_template.h:183]: Reverting to `operator`. Thank you for the heads up

File: torch/CMakeLists.txt [hunk 7]
```diff
@@ -340,13 +340,19 @@ endif()
...
+  string(TOUPPER "$ENV{BUILD_CAFFE2}" BUILD_CAFFE2_STR)
+  if("${BUILD_CAFFE2_STR}" STREQUAL "1" OR "${BUILD_CAFFE2_STR}" STREQUAL "ON" OR "${BUILD_CAFFE2_STR}" STREQUAL "TRUE")
+    string(APPEND CMAKE_CXX_FLAGS " -DBUILD_CAFFE2")
```
Comment (malfet) [inline torch/CMakeLists.txt:346]: `string(APPEND CMAKE_CXX_FLAGS` should not be used anywhere other than at top level file, as it indiscriminately applies this flag to all targets, including third party libraries. 
Please use something more targeted, like  `target_compile_definitions` or `set_source_files_properties`

Regardless, the whole idea that `torch` folder should somehow depend on `BUILD_CAFFE2` options feels wrong. Perhaps this code needs to be moved to `caffe2` or `aten` folder?
Also, it feels semantically wrong that you want to make any code in `torch` folder depend on `caffe2` build options.

**Review by malfet — DISMISSED** [review]
Do not use `string(APPEND CMAKE_CXX_FLAGS` to pass `BUILD_CAFFE2` to the entire project, use either `target_compile_definition`, `set_source_files_properties` or one of `.h.in` files (say `c10/macros/cmake_macros.h.in`)

**Comment by malfet** [comment]
@thiagocrepaldi , can you please elaborate what do you mean by "First, this PR fixes a bug in the cmake build script in which BUILD_CAFFE2 build definition would always be set (even when BUILD_CAFFE2 env var is not defined or set to )/OFF/FALSE"? 

There is `caffe2-linux-xenial-py3.7-gcc5.4` workflow, which generates binaries build with Caffe2 enabled, where all other builds do not do that, you can verify that for yourself by downloading binary artifact for that workflow for example from https://hud.pytorch.org/pytorch/pytorch/commit/fdd12a9f4ce5978f95df82f3ce3e9c3147582a1d

File: torch/CMakeLists.txt (hunk 7 above)
Comment (thiagocrepaldi) [reply torch/CMakeLists.txt]: You are right, thanks!
I have moved the `BUILD_CAFFE2` build option handling to `caffe2/CMakeLists.txt`. That not only have improved code cohesion but also fixed the issue with less lines of code.

The issue I referred to was self-inflicted while I was trying to handle BUILD_CAFFE2 option from `torch/CMakeLists.txt`. In that attempt, BUILD_CAFFE2 didnt read the correct value defined at the main CMakeLists.txt.

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi Thiago Crepaldi FTE , can you please elaborate what do you mean by "First, this PR fixes a bug in the cmake build script in which BUILD_CAFFE2 build definition would always be set (even when BUILD_CAFFE2 env var is not defined or set to )/OFF/FALSE"?
> 
> There is `caffe2-linux-xenial-py3.7-gcc5.4` workflow, which generates binaries build with Caffe2 enabled, where all other builds do not do that, you can verify that for yourself by downloading binary artifact for that workflow for example from https://hud.pytorch.org/pytorch/pytorch/commit/fdd12a9f4ce5978f95df82f3ce3e9c3147582a1d

replied on previous comment. TL;DR I misinterpreted the structure of the build and incorrectly used torch/CMakeLists.txt

**Comment by thiagocrepaldi** [comment]
> @thiagocrepaldi Thiago Crepaldi FTE please respond thoroughly to _all_ the comments and questions from @dzhulgakov and @malfet. They've been generous in using their valuable time to review our contributions, and I don't think you've addressed their comments sufficiently to ask for another round of review yet.

Thanks for the heads up, I will try to summarize below, but LMK if I missed something.

* Nikita's concern was regarding to the build changes on why `BUILD_CAFFE2` definition was being defined globally (all CMake targets).
   *  I have narrowed `BUILD_CAFFE2` definition to `torch` and `torch_python` targets.
      * The former is needed because one of the [ONNX JIT passes(https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/serialization/export.cpp) 
      * The latter because `torch.onnx._CAFFE2_ATEN_FALLBACK` is defined within the python bindings
   * The issue on [why `BUILD_CAFFE2` was always defined for torch target]https://github.com/pytorch/pytorch/blob/master/torch/CMakeLists.txt#L349) didn't really affect PyTorch or Caffe2, but was fixed by the changes mentioned above
* Dmytro and I discussed directly on Slack and his concern was regarding to BC with the attribute renamed from `operator` to `operator_name`. That could be a potential if there are serialized (ONNX) models that are only executed by Caffe2 on master branch (as opposed to re-exporting it).
   * I have acknowledged this possibility and reverted the renaming
   * In fact, the only change to Caffe2 was the Op rename from `onnx::ATen` to `aten::ATen`, which should be OK as per our discussions

In summary, both Caffe2 and PyTorch build changes were minimized to the very least. There were some typos and refactoring to eliminate code dupe

File: test/onnx/expect/TestOperators.test_layer_norm_aten.expect (hunk 1 above)
Comment (dzhulgakov) [reply test/onnx/expect/TestOperators.test_layer_norm_aten.expect:114]: This change still looks suspicious as it's not only one test affected. I'm not familiar with ONNX export shape inference, but it feels like some results became dynamic.

**Review by dzhulgakov — COMMENTED** [review]
Looks good from my side, except for the shape inference turning dynamic

File: test/onnx/expect/TestOperators.test_layer_norm_aten.expect (hunk 1 above)
Comment (jiafatom) [reply test/onnx/expect/TestOperators.test_layer_norm_aten.expect:114]: For onnx domain, shape inference may go to the inference mode, and then `TypeReliable` would be `true`. For aten domain, shape inference is obtained through tracing, so `TypeReliable` would be `false` unless the operator is defined in `static std::unordered_set<std::string> nodeTypeReliableForTracer`. When `TypeReliable` is `false`, it would cause a dynamic shape, that seems the reason here.

Comment (thiagocrepaldi) [reply test/onnx/expect/TestOperators.test_layer_norm_aten.expect]: I have confirmed this by adding `aten::ATen` to the list of reliable types, restoring the original behavior (which might be safer, but less accurate)

**Comment by thiagocrepaldi** [comment]
> Looks good from my side, except for the shape inference turning dynamic

From @jiafatom's analysis:
> For onnx domain, shape inference may go to the inference mode, and then `TypeReliable` would be `true`. For aten domain, shape inference is obtained through tracing, so `TypeReliable` would be `false` unless the operator is defined in `static std::unordered_set<std::string> nodeTypeReliableForTracer`. When `TypeReliable` is `false`, it would cause a dynamic shape, that seems the reason here.

I have confirmed this by adding aten::ATen to the list of reliable types, restoring the original behavior (which might be safer, but less accurate)

File: torch/csrc/jit/passes/onnx/shape_type_inference.cpp [hunk 8]
```diff
@@ -1898,7 +1898,8 @@ static std::unordered_set<std::string> nodeTypeReliableForTracer = {
     "onnx::Cast",
     "onnx::Constant",
     "onnx::Relu",
-    "com.microsoft::Gelu"};
+    "com.microsoft::Gelu",
+    "aten::ATen"};
```
Comment (jiafatom) [inline torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1902]: are we sure whether `aten::ATen` should be in `nodeTypeReliableForTracer` in all scenarios? Maybe we need to add a comment saying that this is for backward compatibility (to achieve consistent result in UT)?

Comment (thiagocrepaldi) [reply torch/csrc/jit/passes/onnx/shape_type_inference.cpp:1902]: The PR is intended to keep back compatibility, so adding `aten::ATen` to `nodeTypeReliableForTracer` is necessary for that intent. It is not desired to have shape inference with different outcomes before and after this PR.

File: torch/testing/_internal/common_quantization.py [hunk 9]
```diff
@@ -294,6 +294,22 @@ def wrapper(*args, **kwargs):
...
+        return fn
+
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        if not torch.onnx._CAFFE2_ATEN_FALLBACK:
+            raise unittest.SkipTest(reason)
+        else:
+            fn(*args, **kwargs)
+    return wrapper
+
```
Comment (malfet) [inline torch/testing/_internal/common_quantization.py:312]: Hmm, this looks like almost an exact duplicate of the same method in `torch/testing/_internal/comon_utils.py`? Why not just import the same method? Or create `_skipIfNotCaffe2WithReason(fn, reason):` in common_utils, import it here and provide another reason?

**Review by malfet — APPROVED** [review]

**Comment by malfet** [comment]
Looks good to me, but please update PR description, as I think 
```
First, this PR fixes a bug in the cmake build script in which BUILD_CAFFE2 build definition would always be set (even when BUILD_CAFFE2 env var is not defined or set to 0/OFF/FALSE for target torch,
```
is not a valid statement, i.e. this PR among other things propagates `BUILD_CAFFE2` flag to `libtorch` in addition to `libtorch_python`)

**Comment by thiagocrepaldi** [comment]
> Looks good to me, but please update PR description, as I think
> 
> ```
> First, this PR fixes a bug in the cmake build script in which BUILD_CAFFE2 build definition would always be set (even when BUILD_CAFFE2 env var is not defined or set to 0/OFF/FALSE for target torch,
> ```
> 
> is not a valid statement, i.e. this PR among other things propagates `BUILD_CAFFE2` flag to `libtorch` in addition to `libtorch_python`)

done, thanks for catching that

**Review by garymm — APPROVED** [review]

**Review by jiafatom — APPROVED** [review]

File: caffe2/contrib/aten/aten_test.py [hunk 10]
```diff
@@ -1,9 +1,3 @@
-
-
-
-
-
-from caffe2.python import core, dyndep
```
Comment (malfet) [inline caffe2/contrib/aten/aten_test.py:6]: This will fail this test, wouldn't it?

**Review by malfet — CHANGES_REQUESTED** [review]

File: caffe2/contrib/aten/aten_test.py [hunk 11]
```diff
@@ -1,9 +1,3 @@
...
 from hypothesis import given
```
Comment (malfet) [inline caffe2/contrib/aten/aten_test.py:1]:
```suggestion
from caffe2.python import core
from hypothesis import given
```

**Review by malfet — APPROVED** [review]
Updated the broken test, will see what will happen

File: caffe2/python/benchmark_generator.py [hunk 12]
```diff
@@ -85,7 +85,7 @@ def make_blob_on_context(blob_name, blob_data, context):
 
     # Handle manual rewrite
     if args.context.upper() == "OPENGL":
-        old_ops = [op for op in predict_net.op]
+        old_ops = list(op for op in predict_net.op)
```
Comment (malfet) [inline caffe2/python/benchmark_generator.py:88]: Just curious, why this change is necesary?

Comment (thiagocrepaldi) [reply caffe2/python/benchmark_generator.py:88]: Not needed, by when I ran `make mypy`, I got a list of issues in files I have changed and this was another occurrence. Maybe my local version was too new or too old, or at least different than the ones PyTorch pipelines use?
